src/sync/sync_session.cpp

```diff
@@ -382,15 +404,17 @@ void SyncSession::nonsync_transact_notify(sync::Session::version_type version)
 
 void SyncSession::revive_if_needed()
 {
-    util::Optional<std::function<SyncLogInHandler>&> log_in_handler;
+    util::Optional<std::function<SyncBindSessionHandler>&> handler;
     {
         std::unique_lock<std::mutex> lock(m_state_mutex);
         if (m_state->revive_if_needed(lock, *this)) {
-            log_in_handler = m_config.log_in_handler;
+            handler = m_config.bind_session_handler;
         }
     }
-    if (log_in_handler)
-        log_in_handler.value()(m_realm_path, m_config);
+    if (handler) {
+        REALM_ASSERT(shared_from_this());
```

Comment:
This may indicate a poor design, or maybe I'm not familiar enough with `std::shared_ptr` idioms. When can this situation occur?